[Feat] 7주차 테스트 코드 작성

jang-namu opened the pull request:
## 📋 이슈 번호
7주차 구현입니다.
track #97 
close #101 
## 🛠 구현 사항
계층별 테스트 코드
TestContainer 이용

## 📚 기타

jang-namu committed:
test: 계층별 테스트 코드 및 TC 추가

NARUBROWN commented on contents/todoListAPI/ch/todo/src/main/java/com/example/todo/config/SecurityConfig.java:
```diff
@@ -0,0 +1,37 @@
+        */
+
+        /*
+        주의! csrf 비활성화.
+        REST API 엔드포인트에 있어서 csrf가 필요하냐는 논쟁거리인듯..
+        ref. https://security.stackexchange.com/questions/166724/should-i-use-csrf-protection-on-rest-api-endpoints
+         */
+        http.csrf().disable();
```
나도 글을 읽어보았는데 이건 나도 좀 알쏭달쏭하네
No Cookie != REST API, No Cookie = No CSRF가 맞는것 같긴 하지만
마지막 답변에서 말했듯이 LocalStorage에 있는 JWT를 탈취하는 것 또한 있을만한 일이니깐.
어떤 방법이 더 나은지에 대해서 확인하기 위해서는 SpringSecurity가 이러한 CSRF에서 어떠한 방어책을 제시하고 있는지 
공부해보면 도움이 될 것 같네!

NARUBROWN commented on contents/todoListAPI/ch/todo/src/test/java/com/example/todo/user/UserControllerTest.java:
```diff
@@ -0,0 +1,62 @@
+//    @Transactional  // 이상하게 Controller에서는 실제 데이터베이스에 작성된다.
+//    @Rollback       // Service Layer에서는 테스트 후 데이터베이스에 남지않는다.
+    void getUser() throws Exception {
+        User mockUser = new User("namu1234@gmail.com", "test123", "namu");
+        userRepository.save(mockUser);
+
+        Long userId = mockUser.getUserId();
+        UserResponseDto mockUserResponseDto = mockUser.toResponseDto();
```
Given-When-Then 패턴에서 Given 부분이 잘못된 것 같아
단위 테스트를 진행할때는 최소한의 Bean만 불러와야 하는데 지금 이 코드에서는 Repository Bean까지 주입되어 있어!
다시 한번 생각해보고 코드를 작성해보자!

jang-namu replied:
엔티티의 기본키 값을 GenertaionType.IDENTITY로 사용하다 보니,
DB에 직접 엔티티를 넣어서 기본키를 생성하려는 목적이었어.

다만, 생각해보니 언급한 말대로 DB에 실제 값을 저장하고 기본키가 제대로 설정되는지, 또는 저장이 잘 되는지 확인하는 것은 컨트롤러 테스트의 책임이 아닌것 같네!

어쩌다보니 책 내용 그대로가 됐지만 다음과 같이 수정할게!

```java
@WebMvcTest
@Testcontainers
@AutoConfigureMockMvc
class UserControllerTest {

    @Autowired
    private MockMvc mockMvc;

    @MockBean
    private UserService userService;

    @Test
    void getUser() throws Exception {
        Long userId = 1L;
        UserResponseDto mockUserResponseDto = UserResponseDto.builder()
                .userId(userId)
                .loginId("namu1234@gmail.com")
                .password("test123")
                .name("namu")
                .build();

        given(userService.getUser(userId)).willReturn(mockUserResponseDto);

        mockMvc.perform(get("/users/{userId}", userId))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.userId").value(userId.intValue()))
                .andExpect(jsonPath("$.loginId").value("namu1234@gmail.com"))
                .andExpect(jsonPath("$.password").value("test123"))
                .andExpect(jsonPath("$.name").value("namu"));
    }
}
```

jang-namu committed:
[Chore] week4/ch -> week5/ch

jang-namu committed:
[Docs] week7 발표자료

jang-namu committed:
test: 테스트 application.properties 분리

jang-namu committed:
test: Repository = testTC profile, Controller/Service = testH2 profile 적용
Controller나 Service의 단위테스트에서는 실제로 DB에 데이터를 CRUD하지 않기 때문에
빠른 테스트를 위해 In-memory DB인 H2를 사용.
Repository의 경우 실제 DB와의 상호작용이 필요하므로 운용환경과 테스트환경을 일치시키기 위해
TC 사용 (다만, 현재 Repostory의 테스트 코드가 적절하진 않다.)

서비스 계층의 테스트의 경우 userId의 생성(GenerationType.IDENTITY)을 위해 DB에 save하는 로직 수정필요

jang-namu commented:
[8bbf45b](https://github.com/inu-appcenter/server-study-15.5/pull/98/commits/8bbf45b9e7e068621432573d924f3208208a80ae)에 마지막에

'서비스 계층의 테스트의 경우 userId의 생성(GenerationType.IDENTITY)을 위해 DB에 save하는 로직 수정필요'
질문이 있습니다. User 엔티티를 반환받아야 하는데, ControllerTest에서 그랬던 것 처럼 Return Object를 빌더나 생성자를 통해 직접 생성할 수 없는 상태입니다.

<img width="933" alt="image" src="https://github.com/inu-appcenter/server-study-15.5/assets/121238128/a0db2551-5934-4025-b539-f759fdb16b83">

Repository가 MockBean이기 때문에 save()를 수행해도 실제 엔티티가 DB에 저장되는 것은 아닙니다.
그 근거로 userId는 여전히 null입니다. (위 테스트코드는 정상적으로 통과합니다.)

어떤식으로 작성해야 현재 코드베이스(User엔티티의 생성자 및 userId 생성규칙)를 유지하며 실제 DB와의 상호작용이 필요하지 않은
단위 테스트를 작성할 수 있을까요??

<img width="1052" alt="스크린샷 2023-12-01 오후 4 19 08" src="https://github.com/inu-appcenter/server-study-15.5/assets/121238128/9c5d7bb3-5b9c-438f-8cb2-59492780efdd">

userId는 null입니다.
